### PR TITLE
FIX: Use batch mode for SQLite3 databases (update migrations)

### PR DESCRIPTION
**Pull request by philipbroistedt**
#287

### CODE REVIEW

File: arbeitszeit_flask/migrations/versions/ddb975a7f24b_delete_tables_message_and_useraction.py [hunk 1]
```diff
@@ -18,7 +18,7 @@
 
 def upgrade():
     op.drop_table("message")
-    op.execute("DROP TYPE IF EXISTS useractiontype CASCADE")
+    #op.execute("DROP TYPE IF EXISTS useractiontype CASCADE")
```
[the remark this answers is not in the document]
Comment (philipbroistedt) [reply arbeitszeit_flask/migrations/versions/ddb975a7f24b_delete_tables_message_and_useraction.py:21]: I took over the above change, but it's untested so far

**Comment by philipbroistedt** [comment]
The requested change was commited, but not pushed. SVN habbits. Furthermore the changes were hiddey by git switch.

**Comment by sloschert** [comment]
I guess there is a bug: 
I switched to your branch, created a fresh postgres database and tried to upgrade it using `flask db upgrade`. It did not succeed, I got the following error:

```
...
INFO  [alembic.runtime.migration] Running upgrade 5e541f042231 -> bc2747c3b268, delete column renewed from table Plan
INFO  [alembic.runtime.migration] Running upgrade bc2747c3b268 -> f5987219755a, Create Accountant user model
INFO  [alembic.runtime.migration] Running upgrade f5987219755a -> c37965cc6adc, Remove .approved property of Plan ORM
INFO  [alembic.runtime.migration] Running upgrade c37965cc6adc -> ddb975a7f24b, delete tables message and useraction
Traceback (most recent call last):
  File "/nix/store/dh3j6xr498kq91g0mciggajbq7l88vwd-python3.10-SQLAlchemy-1.4.41/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1900, in _execute_context
    self.dialect.do_execute(
  File "/nix/store/dh3j6xr498kq91g0mciggajbq7l88vwd-python3.10-SQLAlchemy-1.4.41/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
psycopg2.errors.DependentObjectsStillExist: cannot drop type useractiontype because other objects depend on it
DETAIL:  table user_action column action_type depends on type useractiontype
HINT:  Use DROP ... CASCADE to drop the dependent objects too.


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/nix/store/2awyziyqalakbb1xwr8vxl16qv3fcn6l-python3.10-flask-2.2.2/bin/.flask-wrapped", line 9, in <module>
    sys.exit(main())
  File "/nix/store/2awyziyqalakbb1xwr8vxl16qv3fcn6l-python3.10-flask-2.2.2/lib/python3.10/site-packages/flask/cli.py", line 1047, in main
    cli.main()
  File "/nix/store/pl53qimr6imm2lxa69liajrwv2xpjxsd-python3.10-click-8.1.3/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/nix/store/pl53qimr6imm2lxa69liajrwv2xpjxsd-python3.10-click-8.1.3/lib/python3.10/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/nix/store/pl53qimr6imm2lxa69liajrwv2xpjxsd-python3.10-click-8.1.3/lib/python3.10/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/nix/store/pl53qimr6imm2lxa69liajrwv2xpjxsd-python3.10-click-8.1.3/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/nix/store/pl53qimr6imm2lxa69liajrwv2xpjxsd-python3.10-click-8.1.3/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/nix/store/pl53qimr6imm2lxa69liajrwv2xpjxsd-python3.10-click-8.1.3/lib/python3.10/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/nix/store/2awyziyqalakbb1xwr8vxl16qv3fcn6l-python3.10-flask-2.2.2/lib/python3.10/site-packages/flask/cli.py", line 357, in decorator
    return __ctx.invoke(f, *args, **kwargs)
  File "/nix/store/pl53qimr6imm2lxa69liajrwv2xpjxsd-python3.10-click-8.1.3/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/nix/store/fdqqqv4bj8n4020r4smsgnbwias8wxmd-python3.10-Flask-Migrate-3.1.0/lib/python3.10/site-packages/flask_migrate/cli.py", line 149, in upgrade
    _upgrade(directory, revision, sql, tag, x_arg)
  File "/nix/store/fdqqqv4bj8n4020r4smsgnbwias8wxmd-python3.10-Flask-Migrate-3.1.0/lib/python3.10/site-packages/flask_migrate/__init__.py", line 98, in wrapped
    f(*args, **kwargs)
  File "/nix/store/fdqqqv4bj8n4020r4smsgnbwias8wxmd-python3.10-Flask-Migrate-3.1.0/lib/python3.10/site-packages/flask_migrate/__init__.py", line 185, in upgrade
    command.upgrade(config, revision, sql=sql, tag=tag)
  File "/nix/store/fdnl5kb7y2682g8kmfkf3fc3x9wy6hsx-python3.10-alembic-1.8.1/lib/python3.10/site-packages/alembic/command.py", line 322, in upgrade
    script.run_env()
  File "/nix/store/fdnl5kb7y2682g8kmfkf3fc3x9wy6hsx-python3.10-alembic-1.8.1/lib/python3.10/site-packages/alembic/script/base.py", line 569, in run_env
    util.load_python_file(self.dir, "env.py")
  File "/nix/store/fdnl5kb7y2682g8kmfkf3fc3x9wy6hsx-python3.10-alembic-1.8.1/lib/python3.10/site-packages/alembic/util/pyfiles.py", line 94, in load_python_file
    module = load_module_py(module_id, path)
  File "/nix/store/fdnl5kb7y2682g8kmfkf3fc3x9wy6hsx-python3.10-alembic-1.8.1/lib/python3.10/site-packages/alembic/util/pyfiles.py", line 110, in load_module_py
    spec.loader.exec_module(module)  # type: ignore
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/home/sebas/Coding/Projects/arbeitszeitapp/arbeitszeit_flask/migrations/env.py", line 91, in <module>
    run_migrations_online()
  File "/home/sebas/Coding/Projects/arbeitszeitapp/arbeitszeit_flask/migrations/env.py", line 85, in run_migrations_online
    context.run_migrations()
  File "<string>", line 8, in run_migrations
  File "/nix/store/fdnl5kb7y2682g8kmfkf3fc3x9wy6hsx-python3.10-alembic-1.8.1/lib/python3.10/site-packages/alembic/runtime/environment.py", line 853, in run_migrations
    self.get_context().run_migrations(**kw)
  File "/nix/store/fdnl5kb7y2682g8kmfkf3fc3x9wy6hsx-python3.10-alembic-1.8.1/lib/python3.10/site-packages/alembic/runtime/migration.py", line 623, in run_migrations
    step.migration_fn(**kw)
  File "/home/sebas/Coding/Projects/arbeitszeitapp/arbeitszeit_flask/migrations/versions/ddb975a7f24b_delete_tables_message_and_useraction.py", line 21, in upgrade
    sa.Enum(name="useractiontype").drop(op.get_bind(), checkfirst=False)
  File "/nix/store/dh3j6xr498kq91g0mciggajbq7l88vwd-python3.10-SQLAlchemy-1.4.41/lib/python3.10/site-packages/sqlalchemy/sql/sqltypes.py", line 1181, in drop
    t.drop(bind=bind, checkfirst=checkfirst)
  File "/nix/store/dh3j6xr498kq91g0mciggajbq7l88vwd-python3.10-SQLAlchemy-1.4.41/lib/python3.10/site-packages/sqlalchemy/dialects/postgresql/base.py", line 2062, in drop
    bind._run_ddl_visitor(self.EnumDropper, self, checkfirst=checkfirst)
  File "/nix/store/dh3j6xr498kq91g0mciggajbq7l88vwd-python3.10-SQLAlchemy-1.4.41/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 2211, in _run_ddl_visitor
    visitorcallable(self.dialect, self, **kwargs).traverse_single(element)
  File "/nix/store/dh3j6xr498kq91g0mciggajbq7l88vwd-python3.10-SQLAlchemy-1.4.41/lib/python3.10/site-packages/sqlalchemy/sql/visitors.py", line 524, in traverse_single
    return meth(obj, **kw)
  File "/nix/store/dh3j6xr498kq91g0mciggajbq7l88vwd-python3.10-SQLAlchemy-1.4.41/lib/python3.10/site-packages/sqlalchemy/dialects/postgresql/base.py", line 2104, in visit_enum
    self.connection.execute(DropEnumType(enum))
  File "/nix/store/dh3j6xr498kq91g0mciggajbq7l88vwd-python3.10-SQLAlchemy-1.4.41/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1380, in execute
    return meth(self, multiparams, params, _EMPTY_EXECUTION_OPTS)
  File "/nix/store/dh3j6xr498kq91g0mciggajbq7l88vwd-python3.10-SQLAlchemy-1.4.41/lib/python3.10/site-packages/sqlalchemy/sql/ddl.py", line 80, in _execute_on_connection
    return connection._execute_ddl(
  File "/nix/store/dh3j6xr498kq91g0mciggajbq7l88vwd-python3.10-SQLAlchemy-1.4.41/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1472, in _execute_ddl
    ret = self._execute_context(
  File "/nix/store/dh3j6xr498kq91g0mciggajbq7l88vwd-python3.10-SQLAlchemy-1.4.41/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1943, in _execute_context
    self._handle_dbapi_exception(
  File "/nix/store/dh3j6xr498kq91g0mciggajbq7l88vwd-python3.10-SQLAlchemy-1.4.41/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 2124, in _handle_dbapi_exception
    util.raise_(
  File "/nix/store/dh3j6xr498kq91g0mciggajbq7l88vwd-python3.10-SQLAlchemy-1.4.41/lib/python3.10/site-packages/sqlalchemy/util/compat.py", line 208, in raise_
    raise exception
  File "/nix/store/dh3j6xr498kq91g0mciggajbq7l88vwd-python3.10-SQLAlchemy-1.4.41/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1900, in _execute_context
    self.dialect.do_execute(
  File "/nix/store/dh3j6xr498kq91g0mciggajbq7l88vwd-python3.10-SQLAlchemy-1.4.41/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.InternalError: (psycopg2.errors.DependentObjectsStillExist) cannot drop type useractiontype because other objects depend on it
DETAIL:  table user_action column action_type depends on type useractiontype
HINT:  Use DROP ... CASCADE to drop the dependent objects too.

[SQL: DROP TYPE useractiontype]
(Background on this error at: https://sqlalche.me/e/14/2j85)
```

**Comment by philipbroistedt** [comment]
I have no problem with sqlite

**Comment by sloschert** [comment]
Yes, the problem ocurred wih a postgres database.

**Comment by philipbroistedt** [comment]
ok, I added a check for the current database system.
On my (windows) system it's working with both databases.
Can you please update an check again!?

**Comment by sloschert** [comment]
Now it works when I upgrade, but I get the following error when downgrading:

```
ERROR [flask_migrate] Error: Ambiguous walk
```

I noticed you actually never merged master into your branch `philip/fix-db-migrations-for-sqlite` in the last months, did you? That's why some migration files are missing in your branch.

**Comment by philipbroistedt** [comment]
I didn't merge... with a merge, of the actual master this issue does not apperar any more.
But there are other 'sqlite' issues in the downgrade path. Downgrading was never an issue of this PR

**Comment by sloschert** [comment]
Looks great, thanks for your work. Upgrading and downgrading on a postgres database work fine, also the tests are running. :-)
Would be great if you could merge the current master into your branch (`git merge master`), that makes it easier to review (without that merge I actually get an error when trying to run flask).

**Review by sloschert — APPROVED** [review]

**Pull request merged**